src/pudl/dbt_wrapper.py

```
@@ -1,63 +1,107 @@
+logger = get_logger(__name__)
 
 
 class NodeContext(NamedTuple):
     """Associate a node's *name* with information describing what went wrong."""
 
     name: str
-    context: pd.DataFrame
+    context: str
 
     def pretty_print(self):
         """Nice output for logging to stdout."""
         return f"{self.name}:\n\n{self.context}"
 
 
-def get_context_for_nodes(nodes: list[GenericTestNode]) -> list[NodeContext]:
-    """Post-process node output to figure out failure context.
+class BuildResult(NamedTuple):
+    """Combine overall result with any useful failure context."""
+
+    success: bool
+    failure_contexts: list[NodeContext]
+
+    def format_failure_contexts(self) -> str:
+        """Nice legible output for logs."""
+        return "\n=====\n".join(ctx.pretty_print() for ctx in self.failure_contexts)
+
+
+def __get_failed_nodes(results: RunExecutionResult) -> list[GenericTestNode]:
+    """Get test node output from tests that failed."""
+    return [res.node for res in results if res.status == TestStatus.Fail]
+
+
+def __get_quantile_contexts(
+    nodes: list[GenericTestNode], dbt: dbtRunner, dbt_dir: Path
+) -> list[NodeContext]:
+    """Run debug_quantile_constraints macro for failed quantile constraints.
 
-    For most tests, just run the compiled code in the node.
+    This is a little tricky because the macro output is just logged to
+    stdout, and not stored in the dbt.invoke result. So, for each node, we:
 
-    (TODO 2025-06-09) for weighted quantile tests, run the "debug quantiles" DBT
-    operation.
+    * redirect stdout
+    * run the macro based on node information
+    * parse stdout to get the context
+
+    Also, if a node has multiple parents, we don't know which table to pass into
+    ``debug_quantile_constraints`` so we just skip it.
     """
     contexts = []
+    for node in nodes:
+        parents = node.depends_on.nodes
+        if len(parents) != 1:
+            logger.warning(
```

I didn't find any of these in the `manifest.json` just now but I swear I saw something funny in this way last night... figured it would be good to at least say something instead of just silently ignoring the issue. But also happy to remove this if we feel it's unnecessary.

```python
>>> import json
>>> with open("dbt/target/manifest.json") as f:
...     manifest = json.load(f)
... 
>>> [(name, parents) for name, parents in manifest["parent_map"].items() if "expect_quantile_constraints" in name and len(parents) != 1]
[]
```